README.md
# control_button

Control button with adjustable sections

## About this package

This widget will allow you to create templates for different kind of controllers.

### Variables

|   Variable  |   Type   |   Required  |
| :---------: | :------: | :---------: |
| `externalDiameter` | _double_ | yes |
| `internalDiameter` | _double_ | yes |
| `mainAction` | _Function_ | yes |
| `sections` | _List<Function>_ | yes |
| `elevation` | _double_ | optional |
| `dividerThickness` | _double_ | optional |
| `externalColor` | _Color_ | optional |
| `internalColor` | _Color_ | optional |
| `dividerColor` | _Color_ | optional |
| `shadowDirection` | _Offset_ | optional |
| `sectionOffset` | _FixedAngles_ | optional |

## Images

| Basic | More Sections | Customizable |
| :---: | :-----------: | :----------: |
| <img src="https://flutter.version-one.com/packages/control_button/basic.gif"/> | <img src="https://flutter.version-one.com/packages/control_button/more_buttons.gif"/> | <img src="https://flutter.version-one.com/packages/control_button/custom.gif"/> |

## Usage

To use this plugin, add `control_button` as a [dependency in your pubspec.yaml file](https://flutter.io/platform-plugins/)

``` dart
dependencies:
  control_button: ^0.0.2
```

### Install it 

``` bash
$ flutter pub get
```

### Import it 

``` dart
import 'package:control_button/control_button.dart';
```

## Example

See `example/example.dart`

```dart
ControlButton(
  sectionOffset: FixedAngles.Zero,
  externalDiameter: 300,
  internalDiameter: 120,
  dividerColor: Colors.blue,
  elevation: 2,
  externalColor: Colors.lightBlue[100],
  internalColor: Colors.grey[300],
  mainAction: () => updateState('Selected Center'),
  sections: [
    () => updateState('Selected 1'),
    () => updateState('Selected 2'),
    () => updateState('Selected 3'),
    () => updateState('Selected 4'),
  ],
)
```

CHANGELOG.md
## [0.0.2] - 2020-11-25

* **NEW** now sections have customizable dividers to show its limits.
* **NEW** you can adjust the offset angle of the sections with a `FixedAngle`.
* Updated `example/example.dart`.
* Updated sample images.

## [0.0.1] - 2020-11-24

* First reviewed functional release.


test/control_button_test.dart
import 'package:flutter_test/flutter_test.dart';

import 'package:control_button/control_button.dart';

void main() {
  test('adds one to input values', () {
    final calculator = Calculator();
    expect(calculator.addOne(2), 3);
    expect(calculator.addOne(-7), -6);
    expect(calculator.addOne(0), 1);
    expect(() => calculator.addOne(null), throwsNoSuchMethodError);
  });
}


lib/src/angles.dart
// Copyright (c) 2014, the Dart project authors.  Please see the AUTHORS file
// for details. All rights reserved. Use of this source code is governed by a
// BSD-style license that can be found in the LICENSE file.

import 'dart:math';

/// FixedAngles enum that includes common angles for the ControlButton.
enum FixedAngles {
  Zero,
  Rectangular90,
  Inclined45,
  Inclined120,
  Plane180,
}

/// Angles class provides helpers to convert FixedAngles to degrees or radians.
class Angles {
  final Map<num, double> _angleMap = const {
    0: 0.0,
    1: 1.5708,
    2: 0.785398,
    3: 2.0944,
    4: 3.14159,
  };

  /// fixedToRadian converts an object of class FixedAngles to radians
  double fixedToRadian(FixedAngles angle) {
    return _angleMap[angle.index];
  }

  /// fixedToDegrees converts an object of class FixedAngles to degrees
  double fixedToDegrees(FixedAngles angle) {
    return _angleMap[angle.index] * 180 / pi;
  }
}


lib/src/divider.dart
// Copyright (c) 2014, the Dart project authors.  Please see the AUTHORS file
// for details. All rights reserved. Use of this source code is governed by a
// BSD-style license that can be found in the LICENSE file.

import 'package:flutter/material.dart';

import 'angles.dart';

/// PainterClass class that extends from CustomPainter provides the divisions for the ControlButton.
class PainterClass extends CustomPainter {
  final int sections;
  final FixedAngles angle;
  final Color dividerColor;
  final double dividerThickness;

  /// PainterClass constructor.
  /// You can modify the sections in the ControlButton instance, this will follow along.
  PainterClass({
    this.sections = 2,
    this.dividerThickness = 1.0,
    this.angle = FixedAngles.Zero,
    this.dividerColor = const Color(0x3377E7E7E),
  });

  /// Override method on paint to draw in the canvas.
  @override
  void paint(Canvas canvas, Size size) {
    final pinkPaint = Paint()
      ..style = PaintingStyle.stroke
      ..strokeWidth = dividerThickness
      ..color = dividerColor;

    _drawSections(
      canvas,
      sections,
      size,
      pinkPaint,
      Angles().fixedToRadian(angle),
    );
  }

  @override
  bool shouldRepaint(covariant CustomPainter oldDelegate) => false;

  /// This functions draws the dividers only horizontally and rotates a defined angle while doing it.
  void _drawSections(
      Canvas canvas, int sections, Size size, Paint paint, double angleOffset) {
    double angle = 6.28319 / sections;

    canvas.save();

    if (angleOffset != 0.0) {
      canvas.translate(size.width / 2, size.height / 2);
      canvas.rotate(0.785398);
      canvas.translate(-size.width / 2, -size.height / 2);
    }

    for (int i = 0; i < sections; i++) {
      canvas.translate(size.width / 2, size.height / 2);
      canvas.rotate(angle);
      canvas.translate(-size.width / 2, -size.height / 2);
      canvas.drawLine(
          Offset(0, size.width / 2), Offset(size.width, size.width / 2), paint);
    }

    canvas.restore();
  }
}


lib/src/coordinates_processing.dart
// Copyright (c) 2014, the Dart project authors.  Please see the AUTHORS file
// for details. All rights reserved. Use of this source code is governed by a
// BSD-style license that can be found in the LICENSE file.

import 'dart:math';
import 'package:flutter/foundation.dart';

/// Coordinates class for parsing TapUpDetails elements.
class Coordinates {
  final double posx;
  final double posy;

  /// Coordinates constructor.
  Coordinates({
    @required this.posx,
    @required this.posy,
  });
}

/// CoordinatesProcessiong class.
/// Includes functions related to processing elements of the class Coordinates and returns an angle in degrees.
class CoordinatesProcessing {
  /// This function uses linear ecuation created from two points to map the input number to the output.
  double _numberMap({
    @required double input,
    @required double inputMin,
    @required double inputMax,
    @required double outputMin,
    @required double outputMax,
  }) {
    return ((input - inputMin) *
            (outputMax - outputMin) /
            (inputMax - inputMin)) +
        outputMin;
  }

  /// This function converts an objtect of the class Coordinates into an angle in degrees, from the positive x axis.
  double getAngle(Coordinates coordinates) {
    double position;
    double x = _numberMap(
      input: coordinates.posx,
      inputMax: 300,
      inputMin: 0,
      outputMin: -1,
      outputMax: 1,
    );
    double y = _numberMap(
      input: coordinates.posy,
      inputMax: 0,
      inputMin: 300,
      outputMin: -1,
      outputMax: 1,
    );
    // Throw error if coordinates are (0,0).
    if (x == 0 && y == 0) throw Error();

    // Get angle according to coordinates
    if (x == 0.0 || y == 0.0) {
      (x > 0 && x != 0) ? position = 0 : position = 180;
      (y > 0 && y != 0) ? position = 90 : position = 270;
    } else {
      if (y > 0) {
        position = atan2(y, x) * 180 / pi;
      } else {
        position = 360 + atan2(y, x) * 180 / pi;
      }
    }
    return position;
  }
}


example/example.dart
// Copyright (c) 2014, the Dart project authors.  Please see the AUTHORS file
// for details. All rights reserved. Use of this source code is governed by a
// BSD-style license that can be found in the LICENSE file.

import 'package:flutter/material.dart';

import 'package:control_button/control_button.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Control Button Example',
      theme: ThemeData(
        primarySwatch: Colors.blue,
        visualDensity: VisualDensity.adaptivePlatformDensity,
      ),
      home: MyHomePage(),
    );
  }
}

class MyHomePage extends StatefulWidget {
  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  String text = '';

  void updateState(String showText) {
    setState(() {
      text = showText;
    });
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text('Control Button'),
      ),
      body: Center(
        child: Column(
          mainAxisAlignment: MainAxisAlignment.spaceAround,
          children: [
            ControlButton(
              sectionOffset: FixedAngles.Zero,
              externalDiameter: 300,
              internalDiameter: 120,
              dividerColor: Colors.blue,
              elevation: 2,
              externalColor: Colors.lightBlue[100],
              internalColor: Colors.grey[300],
              mainAction: () => updateState('Selected Center'),
              sections: [
                () => updateState('Selected 1'),
                () => updateState('Selected 2'),
                () => updateState('Selected 3'),
                () => updateState('Selected 4'),
              ],
            ),
            Text(
              text,
              style: TextStyle(
                fontSize: 24,
                color: Theme.of(context).accentColor,
              ),
            ),
          ],
        ),
      ),
    );
  }
}
